File: src/services/email_service.rs
use lettre::{
    AsyncSmtpTransport, AsyncTransport, Message, Tokio1Executor,
    message::{Mailbox, header::ContentType},
    transport::smtp::authentication::Credentials,
};

use crate::{error::AppError, state::EmailConfig};

pub async fn send_reset_email(
    email_config: &EmailConfig,
    to_email: &str,
    reset_link: &str,
) -> Result<(), AppError> {
    let from = Mailbox::new(
        Some(email_config.from_name.clone()),
        email_config
            .from_email
            .parse()
            .map_err(|_| AppError::Internal)?,
    );
    let to = Mailbox::new(None, to_email.parse().map_err(|_| AppError::Internal)?);

    let body = format!(
        "Bạn đã yêu cầu đặt lại mật khẩu.\n\nNhấn link sau để đặt lại: {reset_link}\n\nLink hết hạn sau {} phút.",
        email_config.reset_ttl_minutes
    );

    let email = Message::builder()
        .from(from)
        .to(to)
        .subject("Reset mật khẩu Todo App")
        .header(ContentType::TEXT_PLAIN)
        .body(body)
        .map_err(|_| AppError::Internal)?;

    let creds = Credentials::new(
        email_config.smtp_username.clone(),
        email_config.smtp_password.clone(),
    );

    let mailer = AsyncSmtpTransport::<Tokio1Executor>::relay(&email_config.smtp_host)
        .map_err(|_| AppError::Internal)?
        .credentials(creds)
        .port(email_config.smtp_port)
        .build();

    mailer.send(email).await.map_err(|_| AppError::Internal)?;
    Ok(())
}

File: bun_hono/src/services/emailService.ts
import nodemailer from 'nodemailer';
import { AppError } from '../models/error';
import type { EmailConfig } from '../config/state';

export const sendResetEmail = async (emailConfig: EmailConfig, toEmail: string, resetLink: string) => {
  const transporter = nodemailer.createTransport({
    host: emailConfig.smtpHost,
    port: emailConfig.smtpPort,
    auth: {
      user: emailConfig.smtpUsername,
      pass: emailConfig.smtpPassword,
    },
  });

  const body = `Bạn đã yêu cầu đặt lại mật khẩu.\n\nNhấn link sau để đặt lại: ${resetLink}\n\nLink hết hạn sau ${emailConfig.resetTtlMinutes} phút.`;

  try {
    await transporter.sendMail({
      from: `${emailConfig.fromName} <${emailConfig.fromEmail}>`,
      to: toEmail,
      subject: 'Reset mật khẩu Todo App',
      text: body,
    });
  } catch {
    throw AppError.internal();
  }
};
